// @ionic/angular/dist/providers/nav-controller.js
import { Injectable, Optional } from '@angular/core';
import { Router } from '@angular/router';
var NavController = /** @class */ (function () {
    function NavController(router) {
        this.router = router;
        this.intent = 0 /* Auto */;
        this.animated = true;
        this.stack = [];
    }
    NavController.prototype.goForward = function (url, animated, extras) {
        this.setIntent(1 /* Forward */, animated);
        return this.router.navigateByUrl(url, extras);
    };
    NavController.prototype.goBack = function (url, animated, extras) {
        this.setIntent(2 /* Back */, animated);
        return this.router.navigateByUrl(url, extras);
    };
    NavController.prototype.goRoot = function (url, animated, extras) {
        this.setIntent(3 /* Root */, animated);
        return this.router.navigateByUrl(url, extras);
    };
    NavController.prototype.setIntent = function (intent, animated) {
        this.intent = intent;
        this.animated = (animated === undefined)
            ? intent !== 3 /* Root */
            : animated;
    };
    NavController.prototype.consumeTransition = function () {
        var guessDirection = this.guessDirection();
        var direction = 0;
        var animated = false;
        if (this.intent === 0 /* Auto */) {
            direction = guessDirection;
            animated = direction !== 0;
        }
        else {
            animated = this.animated;
            direction = intentToDirection(this.intent);
        }
        this.intent = 0 /* Auto */;
        this.animated = true;
        return {
            direction: direction,
            animated: animated
        };
    };
    NavController.prototype.guessDirection = function () {
        var index = this.stack.indexOf(document.location.href);
        if (index === -1) {
            this.stack.push(document.location.href);
            return 1;
        }
        else if (index < this.stack.length - 1) {
            this.stack = this.stack.slice(0, index + 1);
            return -1;
        }
        return 0;
    };
    NavController.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    NavController.ctorParameters = function () { return [
        { type: Router, decorators: [{ type: Optional }] }
    ]; };
    return NavController;
}());
export { NavController };
function intentToDirection(intent) {
    switch (intent) {
        case 1 /* Forward */: return 1;
        case 2 /* Back */: return -1;
        default: return 0;
    }
}
